### PR TITLE
rustdoc: Special-case macro lookups less

### PR DESCRIPTION
**Pull request by jyn514**
Previously, rustdoc had 3 fallbacks it used:
1. `resolve_macro_path`
2. `all_macros`
3. `resolve_str_path_error`

Ideally, it would only use `resolve_str_path_error`, to be consistent with other namespaces.
Unfortunately, that doesn't consider macros that aren't defined at module scope;
consider for instance
```rust
{
    struct S;

    macro_rules! mac { () => {} }
    // `mac`'s scope starts here

    /// `mac` <- `resolve_str_path_error` won't see this
   struct Z;

    //`mac`'s scope ends here
}
```

This changes it to only use `all_macros` and `resolve_str_path_error`, and gives
`resolve_str_path_error` precedence over `all_macros` in case there are two macros with the same
name in the same module.

This is a smaller version of https://github.com/rust-lang/rust/pull/91427.

r? @petrochenkov

### CODE REVIEW

File: src/test/rustdoc-ui/intra-doc/macro-rules.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+// check-pass
+#![deny(rustdoc::broken_intra_doc_links)]
+
+#[macro_export]
```
Comment (jyn514) [inline src/test/rustdoc-ui/intra-doc/macro-rules.rs:4]: apparently this `macro_export` isn't necessary for the test to fail :confused: it's unclear to me why *any* macros at all work if something this simple fails.

Comment (jyn514) [reply src/test/rustdoc-ui/intra-doc/macro-rules.rs:4]: ok, this appears to be an actual mcve:
```rust
#![allow(rustdoc::private_intra_doc_links)]
macro_rules! foo {
    () => {};
}


/// [foo!]
pub fn baz() {}
```

**Comment by jyn514** [comment]
> Inconsistency detected by ld.so: dl-tls.c: 493: _dl_allocate_tls_init: Assertion `listp->slotinfo[cnt].gen <= GL(dl_tls_generation)' failed!

oh fun, cc @GuillaumeGomez

**Review by CryptoGodfatherVA2 — APPROVED** [review]

**Comment by petrochenkov** [comment]
The CI failure looks spurious?
@bors r+

**Pull request merged**